openmdao/visualization/meta_model_viewer/meta_model_visualization.py

```diff
@@ -0,0 +1,553 @@
+        self.output_variable = self.output_list.index(new)
+        self._update_all_plots()
+
+    def training_points(self):
+        """
+        Calculate the training points and returns and array containing the position and alpha.
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        array
+            The array of training points and their alpha opacity with respect to the surrogate line
+        """
+        # x_training contains
+        # [x1, x2, x3, x4]
+        # Input Data
+        # Output Data
+        x_training = self.surrogate_comp._training_input
+        y_training = np.squeeze(self.stack_outputs(self.surrogate_comp._training_output), axis=1)
+        output_variable = self.output_list.index(self.output_select.value)
+        data = np.zeros((0, 8))
+
+        # Calculate the limits of each input parameter
+        bounds = [[min(i), max(i)] for i in self.input_data.values()]
+        limits = np.array(bounds)
+        self.limit_range = limits[:, 1] - limits[:, 0]
+
+        # Vertically stack the x/y inputs and then transpose them
+        infos = np.vstack((x_training[:, self.x_index], x_training[:, self.y_index])).transpose()
+        points = x_training.copy()
+        # Set the first two columns of the points array to x/y inputs, respectively
+        points[:, self.x_index] = self.input_point_list[self.x_index]
+        points[:, self.y_index] = self.input_point_list[self.y_index]
+        points = np.divide(points, self.limit_range)
+        tree = cKDTree(points)
+        dist_limit = np.linalg.norm(self.dist_range * self.limit_range)
+        scaled_x0 = np.divide(self.input_point_list, self.limit_range)
+        # Query the nearest neighbors tree for the closest points to the scaled x0 array
+        dists, idx = tree.query(scaled_x0, k=len(x_training), distance_upper_bound=dist_limit)
+        idx = idx[idx != len(x_training)]
+
+        # info contains:
+        # [x_value, y_value, ND-distance, func_value, alpha]
+
+        data = np.zeros((len(idx), 5))
+        for dist_index, i in enumerate(idx):
+            if i != len(x_training):
+                info = np.ones((5))
+                info[0:2] = infos[i, :]
+                info[2] = dists[dist_index] / dist_limit
+                info[3] = y_training[i, output_variable]
+                info[4] = (1. - info[2] / self.dist_range) ** 0.5
+                data[dist_index] = info
+
+        return data
+
+    def stack_outputs(self, outputs_dict):
+        """
+        Stack the values of a dictionary.
+
+        Parameters
+        ----------
+        outputs_dict : dict
+            Dictionary of outputs
+
+        Returns
+        -------
+        array
+            np.stack of values
+        """
+        output_lists_to_stack = []
+        for values in outputs_dict.values():
+            output_lists_to_stack.append(np.asarray(values))
+
+        return np.stack(output_lists_to_stack, axis=-1)
```

Comment:
Another useful tool you can use is `vulture`, which you can pip install.  You can run it on a file (or a group of files) and It can find dead code such as variables or methods that are assigned and not used.

Reply:
This is a great tool. There were few lines I was able to delete.